File: medium/2196_create_binary_tree_from_descriptions/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <unordered_map>
#include <unordered_set>
#include <vector>

class Solution {
public:
  utils::TreeNode *
  createBinaryTree(std::vector<std::vector<int>> &descriptions) {
    std::unordered_map<int, utils::TreeNode *> nodesMap;
    std::unordered_set<int> parentless;

    for (const std::vector<int> &description : descriptions) {
      int parent = description[0];
      int child = description[1];
      bool isLeft = description[2];

      if (nodesMap.find(parent) == nodesMap.end()) {
        nodesMap[parent] = new utils::TreeNode(parent);
        parentless.insert(parent);
      }

      if (nodesMap.find(child) == nodesMap.end())
        nodesMap[child] = new utils::TreeNode(child);
      if (parentless.find(child) != parentless.end())
        parentless.erase(child);

      if (isLeft)
        nodesMap[parent]->left = nodesMap[child];
      else
        nodesMap[parent]->right = nodesMap[child];
    }

    return nodesMap[*parentless.begin()];
  }
};

File: medium/2196_create_binary_tree_from_descriptions/solution.py
from typing import List, Optional

from utils.python.nodes import TreeNode


class Solution:
    def createBinaryTree(
        self, descriptions: List[List[int]]
    ) -> Optional[TreeNode]:
        nodesMap = {}
        parentless = set()

        for parent, child, isLeft in descriptions:
            if parent not in nodesMap:
                nodesMap[parent] = TreeNode(val=parent)
                parentless.add(parent)

            if child not in nodesMap:
                nodesMap[child] = TreeNode(val=child)
            if child in parentless:
                parentless.remove(child)

            if isLeft:
                nodesMap[parent].left = nodesMap[child]
            else:
                nodesMap[parent].right = nodesMap[child]

        return nodesMap[parentless.pop()]
